mlir/include/Quantum/IR/QuantumOps.td

```diff
@@ -194,7 +194,7 @@ def InsertOp : Memory_Op<"insert", [NoMemoryEffect]> {
 // -----
 
 class Gate_Op<string mnemonic, list<Trait> traits = []> :
-        Quantum_Op<mnemonic, traits # [NoMemoryEffect, QuantumGate, Unitary]> {
+        Quantum_Op<mnemonic, traits # [QuantumGate, Unitary]> {
```

Comment:
Do you know why we added NoMemoryEffect?

Reply:
This allows gates to automatically be folded (removed) when their SSA values are unused. This might be the case if you apply a gate on a wire without ever measuring that wire, which makes that gate redundant.

I moved it down to the specific operations so that GlobalPhase can also inherit from the `Gate_Op` class (so that GlobalPhase doesn't have to redefine all the interface methods). GlobalPhase cannot have the NoMemoryEffect attribute since it doesn't necessarily act on any wires.